src/compiler/checker.ts

```diff
@@ -3581,7 +3581,32 @@ namespace ts {
                         // An override clause will take effect for type-only imports and import types, and allows importing the types across formats, regardless of
                         // normal mode restrictions
                         if (isSyncImport && sourceFile.impliedNodeFormat === ModuleKind.ESNext && !getResolutionModeOverrideForClause(overrideClause)) {
-                            error(errorNode, Diagnostics.Module_0_cannot_be_imported_using_this_construct_The_specifier_only_resolves_to_an_ES_module_which_cannot_be_imported_synchronously_Use_dynamic_import_instead, moduleReference);
+                            if (findAncestor(location, isImportEqualsDeclaration)) {
+                                // ImportEquals in a ESM file resolving to another ESM file
+                                error(errorNode, Diagnostics.Module_0_cannot_be_imported_using_this_construct_The_specifier_only_resolves_to_an_ES_module_which_cannot_be_imported_with_require_Use_an_ECMAScript_import_instead, moduleReference);
+                            }
+                            else {
+                                // CJS file resolving to an ESM file
+                                const diag = error(errorNode, Diagnostics.The_current_file_is_a_CommonJS_module_whose_imports_will_produce_require_calls_however_the_referenced_file_is_an_ECMAScript_module_and_cannot_be_imported_with_require_Consider_writing_a_dynamic_import_0_call_instead, moduleReference);
+                                const ext = tryGetExtensionFromPath(currentSourceFile.fileName);
+                                if (ext === Extension.Ts || ext === Extension.Js) {
+                                    const scope = host.getPackageScopeForPath(currentSourceFile.path);
+                                    const targetExt = ext === Extension.Ts ? Extension.Mts : Extension.Mjs;
+                                    if (scope && !scope.packageJsonContent.type) {

```

Comment:
I realized there's another gotcha. If a user has no `package.json`, that's a possible footgun too. A user might try to create a new TS project and not realize that a `package.json` is expected for `nodenext`.

So if you can't find a `package.json`, consider another related diagnostic:

```
To convert this file to an ECMAScript module, change its file extension to '{0}' or create a local package.json file with `{ "type": "module" }`.
```

Reply:
On the other hand, I guess we'll suggest the below diagnostic anyway. Maybe that's good enough. And I see you have a test case for that below.

Reply:
That's fair. No package.json in a Node project is pretty fishy.